File: python/chronos/src/bigdl/chronos/data/utils/scale.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import numpy as np
import pandas as pd

from sklearn.preprocessing import StandardScaler, MaxAbsScaler, MinMaxScaler, RobustScaler


def _standard_scaler_unscale_timeseries_numpy(data, scaler, scaler_index):
    data_unscale = np.zeros(data.shape)
    feature_counter = 0
    for i in scaler_index:
        value_mean = scaler.mean_[i] if scaler.with_mean else 0
        value_scale = scaler.scale_[i] if scaler.with_std else 1
        data_unscale[:, :, feature_counter] = data[:, :, feature_counter] * value_scale + value_mean
        feature_counter += 1
    return data_unscale


def _maxabs_scaler_unscale_timeseries_numpy(data, scaler, scaler_index):
    data_unscale = np.zeros(data.shape)
    feature_counter = 0
    for i in scaler_index:
        value_max_abs = scaler.max_abs_[i]
        data_unscale[:, :, feature_counter] = data[:, :, feature_counter] * value_max_abs
        feature_counter += 1
    return data_unscale


def _minmax_scaler_unscale_timeseries_numpy(data, scaler, scaler_index):
    data_unscale = np.zeros(data.shape)
    feature_counter = 0
    for i in scaler_index:
        value_min = scaler.min_[i]
        value_scale = scaler.scale_[i]
        data_unscale[:, :, feature_counter] = \
            (data[:, :, feature_counter] - value_min) / value_scale
        feature_counter += 1
    return data_unscale


def _robust_scaler_unscale_timeseries_numpy(data, scaler, scaler_index):
    data_unscale = np.zeros(data.shape)
    feature_counter = 0
    for i in scaler_index:
        value_center = 0 if scaler.center_ is None else scaler.center_[i]
        value_scale = 1 if scaler.scale_ is None else scaler.scale_[i]
        data_unscale[:, :, feature_counter] = \
            data[:, :, feature_counter] * value_scale + value_center
        feature_counter += 1
    return data_unscale


UNSCALE_HELPER_MAP = {StandardScaler: _standard_scaler_unscale_timeseries_numpy,
                      MaxAbsScaler: _maxabs_scaler_unscale_timeseries_numpy,
                      MinMaxScaler: _minmax_scaler_unscale_timeseries_numpy,
                      RobustScaler: _robust_scaler_unscale_timeseries_numpy}


def unscale_timeseries_numpy(data, scaler, scaler_index):
    return UNSCALE_HELPER_MAP[type(scaler)](data, scaler, scaler_index)


def _standard_scaler_scale_timeseries_numpy(data, scaler):
    data_scale = np.zeros(data.shape)
    feature_counter = 0
    for i in range(data.shape[1]):
        value_mean = scaler.mean_[i] if scaler.with_mean else 0
        value_scale = scaler.scale_[i] if scaler.with_std else 1
        data_scale[:, i] = (data[:, i] - value_mean) / value_scale
    return data_scale


def _maxabs_scaler_scale_timeseries_numpy(data, scaler):
    data_scale = np.zeros(data.shape)
    for i in range(data.shape[1]):
        value_max_abs = scaler.scale_[i]
        data_scale[:, i] = data[:, i] / value_max_abs
    return data_scale


def _minmax_scaler_scale_timeseries_numpy(data, scaler):
    data_scale = np.zeros(data.shape)
    for i in range(data.shape[1]):
        value_min = scaler.min_[i]
        value_scale = scaler.scale_[i]
        data_scale[:, i] = data[:, i] * value_scale + value_min
    return data_scale


def _robust_scaler_scale_timeseries_numpy(data, scaler):
    data_scale = np.zeros(data.shape)
    for i in range(data.shape[1]):
        value_center = scaler.center_[i] if scaler.with_centering else 0
        value_scale = scaler.scale_[i] if scaler.with_scaling else 1
        data_scale[:, i] = (data[:, i] - value_center) / value_scale
    return data_scale


SCALE_HELPER_MAP = {StandardScaler: _standard_scaler_scale_timeseries_numpy,
                    MaxAbsScaler: _maxabs_scaler_scale_timeseries_numpy,
                    MinMaxScaler: _minmax_scaler_scale_timeseries_numpy,
                    RobustScaler: _robust_scaler_scale_timeseries_numpy}


def scale_timeseries_numpy(data, scaler):
    return SCALE_HELPER_MAP[type(scaler)](data, scaler)

File: scala/dllib/src/main/scala/com/intel/analytics/bigdl/dllib/keras/layers/Scale.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.dllib.keras.layers

import com.intel.analytics.bigdl.dllib.nn.abstractnn.{AbstractModule, IdentityOutputShape}
import com.intel.analytics.bigdl.dllib.tensor.Tensor
import com.intel.analytics.bigdl.dllib.nn.internal.KerasLayer
import com.intel.analytics.bigdl.dllib.utils.Shape
import com.intel.analytics.bigdl.dllib.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.dllib.keras.layers.utils.KerasUtils

import scala.reflect.ClassTag

/**
 * Scale is the combination of CMul and CAdd.
 *
 * Computes the element-wise product of the input and weight,
 * with the shape of the weight "expand" to match the shape of the input.
 * Similarly, perform an expanded bias and perform an element-wise add.
 *
 * When you use this layer as the first layer of a model, you need to provide
 * the argument inputShape (a Single Shape, does not include the batch dimension).
 *
 * Remark: This layer is from Torch and wrapped in Keras style.
 *
 * @param size Size of the weight and bias.
 * @param inputShape A Single Shape, does not include the batch dimension.
 * @tparam T The numeric type of parameter(e.g. weight, bias). Only support float/double now.
 */
class Scale[T: ClassTag](
    val size: Array[Int],
    val inputShape: Shape = null)(implicit ev: TensorNumeric[T])
  extends KerasLayer[Tensor[T], Tensor[T], T](KerasUtils.addBatch(inputShape))
    with IdentityOutputShape {

  override def doBuild(inputShape: Shape): AbstractModule[Tensor[T], Tensor[T], T] = {
    val layer = com.intel.analytics.bigdl.dllib.nn.Scale(size)
    layer.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
  }
}

object Scale {
  def apply[@specialized(Float, Double) T: ClassTag](
    size: Array[Int],
    inputShape: Shape = null)(implicit ev: TensorNumeric[T]): Scale[T] = {
    new Scale[T](size, inputShape)
  }
}
